model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProvider.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.models.map.storage.jpa;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+import javax.persistence.EntityManager;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.map.common.AbstractEntity;
+import org.keycloak.models.map.storage.MapKeycloakTransaction;
+import org.keycloak.models.map.storage.MapStorage;
+import org.keycloak.models.map.storage.MapStorageProvider;
+import org.keycloak.models.map.storage.MapStorageProviderFactory.Flag;
+import org.keycloak.models.map.storage.jpa.client.JpaClientMapKeycloakTransaction;
+
+public class JpaMapStorageProvider implements MapStorageProvider {
+
+    private final JpaMapStorageProviderFactory factory;
+    private final KeycloakSession session;
+    private final EntityManager em;
+    private final Map<Class<?>, Function<EntityManager, MapKeycloakTransaction>> transactionConstructors = new HashMap<>();
```

Comment:
Should this be put into provider factory, perhaps even `static` there? It seems pretty stable and independent, initialized only in the constructor, and there is a reference to `factory` anyway. Ideally there would be a method in the provider factory that would access it rather than a direct access to the field.

Reply:
When there will be `static` map in provider factory, why it should be rather accessed via method in the factory than by direct access to the static field?

Reply:
You can do either, it's matter of the preference. I prefer methods as they have the benefit that if a change in logic is needed later, it can be done at a single place for all callers, but [YMMW](https://dictionary.cambridge.org/us/dictionary/english/ymmv).